### PR TITLE
feat(api): nvim_cmd supports plus ("+cmd", "++opt") flags

### LINKED ISSUE #21687: nvim_parse_cmd doesn't parse ++opt arguments
### Describe the bug

`nvim_parse_cmd` doesn't return command info to handle ++opt.
For example, `nvim_parse_cmd` parses `edit ++ff=unix test.txt` as `edit {file_name}`.
The same is true for `write ++p`.

### Steps to reproduce

minimal.lua
```lua
local cmd = vim.api.nvim_parse_cmd("edit ++ff=unix test.txt", {})
print(vim.inspect(cmd.args)) -- { "++ff=unix test.txt" }
vim.cmd(cmd)
print(vim.fn.bufname("%")) -- test.txt ++ff=unix
```

```
./build/bin/nvim --clean +'luafile minimal.lua'
```

### Expected behavior

nvim_parse_cmd returns cmd to handle ++opt.

```lua
local cmd = vim.api.nvim_parse_cmd("edit ++ff=unix test.txt", {})
vim.cmd(cmd)
print(vim.fn.bufname("%")) -- test.txt
```

### Neovim version (nvim -v)

0.9.0-dev-629+gd89290b45

### Vim (not Nvim) behaves the same?

N/A (nvim_parse_cmd)

### Operating system/version

Ubuntu 22.04

### Terminal name/version

wsltty

### $TERM environment variable

xterm-256color

### Installation

build from repo

### LINKED ISSUE #23782: nvim_cmd+nvim_parse_cmd!=nvim_command for :edit +linenr
### Problem

nvim_parse_cmd not parsing linenr in `:edit +linenr`

### Steps to reproduce

nvim --clean
`:lua vim.api.nvim_command('edit +1 index.html')` -- this works
`:lua vim.api.nvim_cmd(vim.api.nvim_parse_cmd('edit +1 index.html',{}),{})` -- this opens "+1 index.html”

### Expected behavior

Both command should open index.html, first line.

### Neovim version (nvim -v)

v0.9.0

### Vim (not Nvim) behaves the same?

irrelevant 

### Operating system/version

archlinuxarm

### Terminal name/version

Termux

### $TERM environment variable

xterm-256color

### Installation

pacman

### PR DESCRIPTION
**Pull request by glepnir**
Problem: currently nvim_cmd can not handle plus flags correct.

Solution: add plus flags support.

Fix #21687
Fix #23782

### CODE REVIEW

File: test/functional/api/vim_spec.lua [hunk 1]
```diff
@@ -5205,4 +5205,172 @@ describe('API', function()
...
+
+    it('handles ++ff correctly', function()
+      exec_lua [[
+        vim.cmd(vim.api.nvim_parse_cmd('edit ++ff=mac test.txt', {}))
+      ]]
+      eq('mac', api.nvim_get_option_value('fileformat', {}))
+      eq('test.txt', fn.fnamemodify(api.nvim_buf_get_name(0), ':t'))
+    end)
+
+    it('handles ++fileformat correctly', function()
```
Comment (justinmk) [inline test/functional/api/vim_spec.lua:5254]: these tests can be combined.

please look for similar opportunities in the other tests. avoid verbose, single-purpose tests when multiple related things can be covered in a single test.

**Comment by glepnir** [comment]
there has `getargopt` a static function can be used .  I've noticed it before. But I want to try to see if my current thinking is correct. It looks OK now. I'll try to simplify

File: src/nvim/api/command.c [hunk 2]
```diff
@@ -633,6 +633,24 @@ String nvim_cmd(uint64_t channel_id, Dict(cmd) *cmd, Dict(cmd_opts) *opts, Arena
...
+        goto end;
+      }
+    }
+  }
+
+  // Check for "+command" argument.
+  if ((ea.argt & EX_CMDARG) && !ea.usefilter) {
+    ea.do_ecmd_cmd = getargcmd(&ea.arg);
+  }
+
```
Comment (zeertzjq) [inline src/nvim/api/command.c:653]: Should this be done in `execute_cmd()` instead?

Comment (zeertzjq) [reply src/nvim/api/command.c:653]: Oh, it shouldn't.

File: src/nvim/api/command.c [hunk 3]
```diff
@@ -633,6 +633,24 @@ String nvim_cmd(uint64_t channel_id, Dict(cmd) *cmd, Dict(cmd_opts) *opts, Arena
...
+        api_set_error(err, kErrorTypeValidation, "Invalid argument");
+        goto end;
+      }
+    }
+  }
+
+  // Check for "+command" argument.
+  if ((ea.argt & EX_CMDARG) && !ea.usefilter) {
+    ea.do_ecmd_cmd = getargcmd(&ea.arg);
+  }
```
Comment (famiu) [inline src/nvim/api/command.c:652]: Would it be possible to move this code from here and `do_one_cmd` to `execute_cmd0` so there isn't any code duplication?

Comment (glepnir) [reply src/nvim/api/command.c]: buffer command requires additional processing, so it is not appropriate to move it to execute_cmd0. oldtest `test_buffer` will failed.

File: src/nvim/api/command.c [hunk 4]
```diff
@@ -634,6 +634,21 @@ String nvim_cmd(uint64_t channel_id, Dict(cmd) *cmd, Dict(cmd_opts) *opts, Arena
   build_cmdline_str(&cmdline, &ea, &cmdinfo, args);
   ea.cmdlinep = &cmdline;
 
+  // Check for "++opt=val" argument.
+  if (ea.argt & EX_ARGOPT) {
+    while (ea.arg[0] == '+' && ea.arg[1] == '+') {
+      if (getargopt(&ea) == FAIL && !is_cmd_ni(ea.cmdidx)) {
+        api_set_error(err, kErrorTypeValidation, "Invalid argument");
```
Comment (justinmk) [inline src/nvim/api/command.c:641]: can we use `api_err_invalid` or `VALIDATE_S` so that the invalid arg is mentioned in he message?

File: test/functional/api/vim_spec.lua [hunk 5]
```diff
@@ -5507,4 +5507,156 @@ describe('API', function()
 
     n.assert_alive()
   end)
+
+  describe('nvim_cmd with plus flags', function()
+    it('handles +flags correctly', function()
```
Comment (justinmk) [inline test/functional/api/vim_spec.lua:5512]: these it() cases can live in the existing describe() for nvim_cmd https://github.com/neovim/neovim/blob/91e116f3a6719e9dc7652f4e0a4b1760dfbf1af0/test/functional/api/vim_spec.lua#L4804

**Review by justinmk — APPROVED** [review]
Code is simplified now, thanks. And the tests look thorough. Don't we need a test(s) that check the actual structure returned by nvim_parse_cmd ? There are none currently. Need at least 1 or 2, don't need to cover every case you have added here.

**Review by famiu — APPROVED** [review]

File: test/functional/api/vim_spec.lua [hunk 6]
```diff
@@ -5233,6 +5233,167 @@ describe('API', function()
...
+      -- Write a file for testing +flags
+      t.write_file('testfile', 'Line 1\nLine 2\nLine 3', false, false)
+
+      -- Test + command (go to the last line)
+      local result = exec_lua([[
+        local parsed = vim.api.nvim_parse_cmd('edit + testfile', {})
+        vim.cmd(parsed)
+        return { vim.fn.line('.'), parsed.args }
+      ]])
+      eq({ 3, { '+ testfile' } }, result)
```
Comment (justinmk) [inline test/functional/api/vim_spec.lua:5247]: other than `.args`, are the +/++ things represented in any other field? like a "mods" field or something?

Comment (glepnir) [reply test/functional/api/vim_spec.lua]: okay no test case used command modifier.. I have add a case for  mods and `parsed.cmd`
